subsys/llext/llext_load.c

```diff
@@ -112,56 +139,42 @@ static int llext_load_elf_data(struct llext_loader *ldr, struct llext *ext)
  */
 static int llext_find_tables(struct llext_loader *ldr)
 {
-	int sect_cnt, i, ret;
-	size_t pos;
-	elf_shdr_t shdr;
+	int sect_cnt, i;
 
-	ldr->sects[LLEXT_MEM_SHSTRTAB] =
-		ldr->sects[LLEXT_MEM_STRTAB] =
-		ldr->sects[LLEXT_MEM_SYMTAB] = (elf_shdr_t){0};
+	memset(ldr->sects, 0, sizeof(ldr->sects));
 
 	/* Find symbol and string tables */
-	for (i = 0, sect_cnt = 0, pos = ldr->hdr.e_shoff;
-	     i < ldr->hdr.e_shnum && sect_cnt < 3;
-	     i++, pos += ldr->hdr.e_shentsize) {
-		ret = llext_seek(ldr, pos);
-		if (ret != 0) {
-			LOG_ERR("failed seeking to position %zu\n", pos);
-			return ret;
-		}
-
-		ret = llext_read(ldr, &shdr, sizeof(elf_shdr_t));
-		if (ret != 0) {
-			LOG_ERR("failed reading section header at position %zu\n", pos);
-			return ret;
-		}
-
-		LOG_DBG("section %d at %zx: name %d, type %d, flags %zx, "
-			"ofs %zx, addr %zx, size %zd",
-			i, pos,
-			shdr.sh_name,
-			shdr.sh_type,
-			(size_t)shdr.sh_flags,
-			(size_t)shdr.sh_offset,
-			(size_t)shdr.sh_addr,
-			(size_t)shdr.sh_size);
-
-		switch (shdr.sh_type) {
+	for (i = 0, sect_cnt = 0; i < ldr->sect_cnt; ++i) {
+		elf_shdr_t *shdr = ldr->sect_hdrs + i;
+
+		LOG_DBG("section %d at 0x%zx: name %d, type %d, flags 0x%zx, "
+			"addr 0x%zx, size %zd, link %d, info %d",
+			i,
+			(size_t)shdr->sh_offset,
+			shdr->sh_name,
+			shdr->sh_type,
+			(size_t)shdr->sh_flags,
+			(size_t)shdr->sh_addr,
+			(size_t)shdr->sh_size,
+			shdr->sh_link,
+			shdr->sh_info);
+
+		switch (shdr->sh_type) {
 		case SHT_SYMTAB:
 		case SHT_DYNSYM:
 			LOG_DBG("symtab at %d", i);
-			ldr->sects[LLEXT_MEM_SYMTAB] = shdr;
+			ldr->sects[LLEXT_MEM_SYMTAB] = *shdr;
 			ldr->sect_map[i] = LLEXT_MEM_SYMTAB;
 			sect_cnt++;
 			break;
 		case SHT_STRTAB:
 			if (ldr->hdr.e_shstrndx == i) {
 				LOG_DBG("shstrtab at %d", i);
-				ldr->sects[LLEXT_MEM_SHSTRTAB] = shdr;
+				ldr->sects[LLEXT_MEM_SHSTRTAB] = *shdr;
```

Comment:
Do we need to copy the section headers here or can we directly reference them?

Reply:
See other comment - yes, we do need the copies for the other non-string sections.